### ISSUE TITLE
Squash codeclimate image

### ISSUE THREAD

**Issue by fabry00** [codeclimate/codeclimate#813]
Hi guys,

I'm using codeclimate docker image in gitlab-ci to analyse my code automatically.
It works great. The only problem that I'm experiencing is the time to complete the task.

I've noticed that my gitlab-runner takes a lot of time (15 minutes) to download end decompress all the layers of the codeclimate/codeclimate docker image:

```
D, [2018-02-19T07:59:42.735872 #1] DEBUG -- : engine stderr: Unable to find image 'codeclimate/codeclimate-structure:latest' locally
D, [2018-02-19T07:59:44.140599 #1] DEBUG -- : engine stderr: latest: Pulling from codeclimate/codeclimate-structure
D, [2018-02-19T07:59:44.141209 #1] DEBUG -- : engine stderr: e7bb522d92ff: Pulling fs layer
D, [2018-02-19T07:59:44.141559 #1] DEBUG -- : engine stderr: a3ed95caeb02: Pulling fs layer
D, [2018-02-19T07:59:44.141874 #1] DEBUG -- : engine stderr: 3fe6cc8b54ad: Pulling fs layer
D, [2018-02-19T07:59:44.142185 #1] DEBUG -- : engine stderr: e5f808a44a18: Pulling fs layer
D, [2018-02-19T07:59:44.142490 #1] DEBUG -- : engine stderr: 2134e4618061: Pulling fs layer
D, [2018-02-19T07:59:44.142771 #1] DEBUG -- : engine stderr: e77e198bbb98: Pulling fs layer
D, [2018-02-19T07:59:44.143052 #1] DEBUG -- : engine stderr: 7755344de10e: Pulling fs layer
D, [2018-02-19T07:59:44.143321 #1] DEBUG -- : engine stderr: 937405bec487: Pulling fs layer
D, [2018-02-19T07:59:44.143610 #1] DEBUG -- : engine stderr: e3ee7bbe8387: Pulling fs layer
D, [2018-02-19T07:59:44.143905 #1] DEBUG -- : engine stderr: 6f2dc2d09bc0: Pulling fs layer
D, [2018-02-19T07:59:44.144185 #1] DEBUG -- : engine stderr: b7a149234d3d: Pulling fs layer
D, [2018-02-19T07:59:44.144470 #1] DEBUG -- : engine stderr: 8ed6ef86f959: Pulling fs layer
D, [2018-02-19T07:59:44.144748 #1] DEBUG -- : engine stderr: 58455758e9b0: Pulling fs layer
D, [2018-02-19T07:59:44.145044 #1] DEBUG -- : engine stderr: 4f05907d77f5: Pulling fs layer
```

I was wondering if there is a workaround to speedup this process.

What I've tried is this ( [source](https://forums.docker.com/t/how-to-flatten-an-image-with-127-parents/1600/3)): 

```
docker pull codeclimate/codeclimate
docker run --name mycontainer codeclimate/codeclimate 
docker export --output=myimage.tar mycontainer
cat myimage.tar | docker import myuser/myimage:1.0.0
```
Now the download phase is very fast but I've got the following error:
```

docker run --env CODECLIMATE_DEBUG=1 --env CODECLIMATE_CODE="$PWD/src" --volume "$PWD/src":/code --volume /var/run/docker.sock:/var/run/docker.sock --volume /tmp/cc:/tmp/cc  myuser/myimage:1.0.0 validate-config
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"validate-config\": executable file not found in $PATH": unknown.
time="2018-02-19T09:11:41Z" level=error msg="error waiting for container: context canceled"
ERROR: Job failed: exit code 127
```

Server info:

```
Linux icvs019 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux


Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:11:19 2017
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:09:53 2017
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by dhekimian** [comment]
Do you have the `overlay2` driver enabled?
https://docs.gitlab.com/ce/ci/docker/using_docker_build.html#using-the-overlayfs-driver

or in your `.gitlab-ci.yml`:
```
codequality:
  variables:
    DOCKER_DRIVER: overlay2
```

**Comment by wfleming** [comment]
I'm going to close this. We've evaluated squashing this image in the past and decided against it: the image is already fairly small (a bit over 100MB), and the Dockerfile isn't doing any large deletes in later layers, so squashing wouldn't make it much smaller. Squashing also removes the usefulness of cached older layers: right now when we release CLI updates, most users who upgrade will only need to pull one or two slightly different layers, not the entire image all over again, but if we released squashed images the entire image would need to be pulled every time.

Given David's comment, it sounds like the performance is more of an issue with GitLab's CI environment, and there may be configuration changes that will help there.
